sql: virtual index range scans cause panic

Running queries with `pgadmin` currently causes a node crash, with the following stack trace:

```
panic: mutation disallowed

goroutine 156 [running]:
github.com/cockroachdb/cockroach/pkg/sql/opt/constraint.(*Spans).Truncate(0xc0020520c0, 0x1)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/constraint/spans.go:81 +0x11c
github.com/cockroachdb/cockroach/pkg/sql.virtualDefEntry.makeConstrainedRowsGenerator.func1(0x9437580, 0xc001fe6860, 0x2, 0x1)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:604 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func2(0x94d74e0, 0xc001fec480, 0xc00188a1e0, 0xc0009be690, 0xc001fe6860)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:115 +0x109
created by github.com/cockroachdb/cockroach/pkg/sql.setupGenerator
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:104 +0xf7
```

This ticket tracks fixing the immediate bug, as well as making sure that panics from the virtual table row pusher don't crash the server and are instead forwarded to the main SQL execution goroutine.

So far, I don't think this is reproducible on 20.1. Here is a query that will reproduce the panic on 20.2:

```
demo@127.0.0.1:64504/movr> select * from pg_catalog.pg_class where oid > 100 and oid < 200;
panic: mutation disallowed

goroutine 5825 [running]:
github.com/cockroachdb/cockroach/pkg/sql/opt/constraint.(*Spans).Truncate(0xc001c74ac0, 0x1)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/constraint/spans.go:81 +0x11c
github.com/cockroachdb/cockroach/pkg/sql.virtualDefEntry.makeConstrainedRowsGenerator.func1(0x9437580, 0xc0021ae960, 0x2, 0x1)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:604 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func2(0x94d74e0, 0xc00312dfc0, 0xc001d17560, 0xc0031383c0, 0xc0021ae960)
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:115 +0x109
created by github.com/cockroachdb/cockroach/pkg/sql.setupGenerator
	/Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:104 +0xf7
```

--- a/pkg/sql/logictest/testdata/logic_test/pg_catalog
+++ b/pkg/sql/logictest/testdata/logic_test/pg_catalog
@@ -2769,3 +2769,14 @@ FROM
 WHERE
   n.nspname != 'pg_toast'
   AND (t.typrelid = 0 OR (SELECT c.relkind = 'c' FROM pg_catalog.pg_class AS c WHERE c.oid = t.typrelid))
+
+# Regression test for range constraints on virtual index scans. (#56440)
+query O
+SELECT oid FROM pg_type WHERE oid IN (19,20,24) ORDER BY oid
+----
+19
+20
+24
+
+statement ok
+SELECT * FROM pg_class WHERE oid = 10 OR oid BETWEEN 20 AND 30 OR oid = 40

--- a/pkg/sql/virtual_schema.go
+++ b/pkg/sql/virtual_schema.go
@@ -538,24 +538,26 @@ func (e virtualDefEntry) makeConstrainedRowsGenerator(
 	def := e.virtualDef.(virtualSchemaTable)
 	return func(pusher rowPusher) error {
 		var span constraint.Span
-		addRowIfPassesFilter := func(datums ...tree.Datum) error {
-			for i, id := range index.ColumnIDs {
-				indexKeyDatums[i] = datums[columnIdxMap[id]]
-			}
-			// Construct a single key span out of the current row, so that
-			// we can test it for containment within the constraint span of the
-			// filter that we're applying. The results of this containment check
-			// will tell us whether or not to let the current row pass the filter.
-			key := constraint.MakeCompositeKey(indexKeyDatums...)
-			span.Init(key, constraint.IncludeBoundary, key, constraint.IncludeBoundary)
-			var err error
-			if idxConstraint.ContainsSpan(p.EvalContext(), &span) {
-				if err := e.validateRow(datums, columns); err != nil {
-					return err
+		addRowIfPassesFilter := func(idxConstraint *constraint.Constraint) func(datums ...tree.Datum) error {
+			return func(datums ...tree.Datum) error {
+				for i, id := range index.ColumnIDs {
+					indexKeyDatums[i] = datums[columnIdxMap[id]]
 				}
-				return pusher.pushRow(datums...)
+				// Construct a single key span out of the current row, so that
+				// we can test it for containment within the constraint span of the
+				// filter that we're applying. The results of this containment check
+				// will tell us whether or not to let the current row pass the filter.
+				key := constraint.MakeCompositeKey(indexKeyDatums...)
+				span.Init(key, constraint.IncludeBoundary, key, constraint.IncludeBoundary)
+				var err error
+				if idxConstraint.ContainsSpan(p.EvalContext(), &span) {
+					if err := e.validateRow(datums, columns); err != nil {
+						return err
+					}
+					return pusher.pushRow(datums...)
+				}
+				return err
 			}
-			return err
 		}
 
 		// We have a virtual index with a constraint. Run the constrained
@@ -563,13 +565,9 @@ func (e virtualDefEntry) makeConstrainedRowsGenerator(
 		// index for a given span, we exit the loop early and run a "full scan"
 		// over the virtual table, filtering the output using the remaining
 		// spans.
-		// N.B. we count down in this loop so that, if we have to give up half
-		// way through, we can easily truncate the spans we already processed
-		// from the end and use them as a filter for the remaining rows of the
-		// table.
-		currentConstraint := idxConstraint.Spans.Count() - 1
-		for ; currentConstraint >= 0; currentConstraint-- {
-			span := idxConstraint.Spans.Get(currentConstraint)
+		var currentSpan int
+		for ; currentSpan < idxConstraint.Spans.Count(); currentSpan++ {
+			span := idxConstraint.Spans.Get(currentSpan)
 			if span.StartKey().Length() > 1 {
 				return errors.AssertionFailedf(
 					"programming error: can't push down composite constraints into vtables")
@@ -584,7 +582,7 @@ func (e virtualDefEntry) makeConstrainedRowsGenerator(
 			// For each span, run the index's populate method, constrained to the
 			// constraint span's value.
 			found, err := virtualIndex.populate(ctx, constraintDatum, p, dbDesc,
-				addRowIfPassesFilter)
+				addRowIfPassesFilter(idxConstraint))
 			if err != nil {
 				return err
 			}
@@ -594,15 +592,21 @@ func (e virtualDefEntry) makeConstrainedRowsGenerator(
 				break
 			}
 		}
-		if currentConstraint < 0 {
+		if currentSpan == idxConstraint.Spans.Count() {
 			// We successfully processed all constraints, so we can leave now.
 			return nil
 		}
 
 		// Fall back to a full scan of the table, using the remaining filters
 		// that weren't able to be used as constraints.
-		idxConstraint.Spans.Truncate(currentConstraint + 1)
-		return def.populate(ctx, p, dbDesc, addRowIfPassesFilter)
+		newConstraint := *idxConstraint
+		newConstraint.Spans = constraint.Spans{}
+		nSpans := idxConstraint.Spans.Count() - currentSpan
+		newConstraint.Spans.Alloc(nSpans)
+		for ; currentSpan < idxConstraint.Spans.Count(); currentSpan++ {
+			newConstraint.Spans.Append(idxConstraint.Spans.Get(currentSpan))
+		}
+		return def.populate(ctx, p, dbDesc, addRowIfPassesFilter(&newConstraint))
 	}
 }
 

